### PR TITLE
Add link validation (fix #122)

### LINKED ISSUE #122: Add Link schema validation methods
We need [schema validation](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/common.py#L47-L69) for Links similar to what we do for [Layouts](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/layout.py#L206-L275), [Steps](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/layout.py#L337-L381) and [Inspections](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/layout.py#L432-L460).

### PR DESCRIPTION
**Pull request by lukpueh**
This PR adds validation methods to `link.LinkSignable` that check the format of the corresponding fields. (closes #122)

By calling `self.validate()` on a`LinkSignable` object  all fields' formats are checked.

This PR also
- adds validation to `LinkSignable`'s constructor,
- adapts existing unit tests where schema validation was less strict, and
- adds new unit tests for the new validation methods.

### CODE REVIEW

File: in_toto/models/link.py [hunk 1]
```diff
@@ -141,3 +141,59 @@ def __init__(self, **kwargs):
     self.byproducts = kwargs.get("byproducts", {})
     self.command = kwargs.get("command", [])
     self.return_value = kwargs.get("return_value", None)
+
+    self.validate()
+
```
Comment (SantiagoTorres) [inline in_toto/models/link.py:152]: I think you need another space here, or less space on the ones below

File: in_toto/models/link.py [hunk 2]
```diff
@@ -141,3 +141,59 @@ def __init__(self, **kwargs):
...
+      raise securesystemslib.exceptions.FormatError(
+          "Invalid Link: field `_type` must be set to 'link', got: {}"
+          .format(self._type))
+
+
+  def _validate_materials(self):
+    """Private method to check that `materials` is a `dict` of `HASHDICTs`."""
+    if not isinstance(self.materials, dict):
+      raise securesystemslib.exceptions.FormatError(
+          "Invalid Link: field `materials` must be of type list, got: {}"
```
Comment (SantiagoTorres) [inline in_toto/models/link.py:166]: I wonder if we can coalesce this to a format string (probably with our own exception class) and just do:
```
raise InvalidLinkException(field='materials', got=type(self.materials))
```

Here and elsewhere.

File: test/models/test_link.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python
+"""
+<Program Name>
+  test_link.signed.py
+
+<Author>
+  Lukas Puehringer <lukas.puehringer@nyu.edu>
+
+<Started>
+  Sep 28, 2011
```
Comment (SantiagoTorres) [inline test/models/test_link.py]: I don't think this is the right date

File: test/models/test_link.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    with self.assertRaises(FormatError):
+      test_link.signed.validate()
+
+
+  def test_validate_return_value(self):
+    """Test `return_value` field. Must be either an `int` or `None` """
+    test_link = Link()
+
+    # Good return_value 1
+    test_link.signed.retunr_value = 1
```
Comment (SantiagoTorres) [inline test/models/test_link.py]: I don't think this is checking what it should check.

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Four really minor details. I think once these are addressed this is ready to merge.

File: in_toto/models/link.py (hunk 2 above)
Comment (lukpueh) [reply in_toto/models/link.py:166]: We could and should, but let's do this elsewhere. Care for a "revisit exception taxonomy" ticket?

**Comment by lukpueh** [comment]
Thanks for the comments, @SantiagoTorres. Please merge when ready!

File: in_toto/models/link.py (hunk 2 above)
Comment (lukpueh) [reply in_toto/models/link.py:166]: FYI: https://github.com/in-toto/in-toto/issues/126

**Review by SantiagoTorres — APPROVED** [review]

**Pull request merged**